### PR TITLE
Show better error when h/1 is used on callback

### LINKED ISSUE #6204: Consider printing something when h/1 helper is used with a callback
### Current behavior

Right now when the `h/1` helper is used in `iex` with a callback, a regular "no function" message is printed. While this is technically correct, it is not useful for the user. 

```
iex(1)> h GenServer.init
No documentation for GenServer.init was found
```

### Expected behavior

I suggest printing something like:
```
iex(1)> h GenServer.init
No documentation for function GenServer.init was found, but there exists a callback
with the same name. You can view callback documentations with b/1 helper.
```

### PR DESCRIPTION
**Pull request by uesteibar**
closes #6204 

## Current behaviour
When calling `h/1` on a `callback`
```
iex(1)> h GenServer.init
No documentation for GenServer.init was found
```

## After this change
```
iex(1)> h GenServer.init
No documentation for function GenServer.init was found, but there is a callback
with the same name.
You can view callback documentations with the b/1 helper.
```

when requesting specific arity
```
iex(1)> h GenServer.init/1
No documentation for function GenServer.init/1 was found, but there is a callback
with the same name.
You can view callback documentations with the b/1 helper.
```

Any feedback is welcome, thanks! :)

### CODE REVIEW

File: lib/iex/lib/iex/introspection.ex [hunk 1]
```diff
@@ -115,13 +123,42 @@ defmodule IEx.Introspection do
...
+    Enum.find(docs, fn ({{fun_name, _}, _, _, _}) -> fun_name == fun end)
+  end
+  def find_callback_doc(docs, fun, arity) do
+    Enum.find(docs, fn ({{fun_name, fun_arity}, _, _, _}) ->
+      fun_name == fun && fun_arity == arity
+    end)
+  end
+
+  def handle_has_callback(nil), do: false
+  def handle_has_callback(_),   do: true
```
Comment (michalmuskala) [inline lib/iex/lib/iex/introspection.ex]: We can get rid of this by using `Enum.any?` instead of `Enum.find`

File: lib/iex/lib/iex/introspection.ex [hunk 2]
```diff
@@ -115,13 +123,42 @@ defmodule IEx.Introspection do
...
+  end
+  defp has_callback?(for, fun, arity) do
+    for
+    |> Code.get_docs(:callback_docs)
+    |> find_callback_doc(fun, arity)
+    |> handle_has_callback()
+  end
+
+  def find_callback_doc(docs, fun) do
+    Enum.find(docs, fn ({{fun_name, _}, _, _, _}) -> fun_name == fun end)
```
Comment (michalmuskala) [inline lib/iex/lib/iex/introspection.ex]: We can use `match?/2` here (and below as well)

File: lib/iex/lib/iex/introspection.ex [hunk 3]
```diff
@@ -70,7 +72,11 @@ defmodule IEx.Introspection do
         h(mod, fun, arity)
       end
 
-      if result != [], do: :ok, else: :not_found
+      cond do
+        result != []            -> :ok
+        has_callback?(mod, fun) -> :behaviour_found
+        true                    -> :not_found
```
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: I think we are avoiding the additional spaces to indent things. So maybe put them in a new line
```elixir
result != []
  -> :ok
```

Comment (uesteibar) [inline lib/iex/lib/iex/introspection.ex]: @eksperimental done! :)

File: lib/iex/lib/iex/introspection.ex [hunk 4]
```diff
@@ -115,13 +126,42 @@ defmodule IEx.Introspection do
...
+        else
+          :not_found
+        end
       end
     else
       :no_docs
     end
   end
 
+  defp has_callback?(for, fun) do
```
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: I was shocked to see `for |>`
then i realized `for` is a variable. I think it will be wise to choose a different name

File: lib/iex/lib/iex/introspection.ex [hunk 5]
```diff
@@ -115,13 +126,42 @@ defmodule IEx.Introspection do
...
+    |> Code.get_docs(:callback_docs)
+    |> find_callback_doc(fun, arity)
+    |> handle_has_callback()
+  end
+
+  def find_callback_doc(docs, fun) do
+    Enum.find(docs, fn ({{fun_name, _}, _, _, _}) -> fun_name == fun end)
+  end
+  def find_callback_doc(docs, fun, arity) do
+    Enum.find(docs, fn ({{fun_name, fun_arity}, _, _, _}) ->
```
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: IIRCC we are no longer using parenthesis in anonymous functions

File: lib/iex/lib/iex/introspection.ex [hunk 6]
```diff
@@ -115,13 +126,42 @@ defmodule IEx.Introspection do
...
+      fun_name == fun && fun_arity == arity
+    end)
+  end
+
+  def handle_has_callback(nil), do: false
+  def handle_has_callback(_),   do: true
```
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: remove indentation

File: lib/iex/lib/iex/introspection.ex [hunk 7]
```diff
@@ -442,6 +482,13 @@ defmodule IEx.Introspection do
   defp no_types(for), do: no(for, "type information")
   defp no_docs(for),  do: no(for, "documentation")
```
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: since we are touching these lines, we can also remove indentation here

File: lib/iex/lib/iex/introspection.ex (hunk 4 above)
Comment (uesteibar) [inline lib/iex/lib/iex/introspection.ex]: I used `for` to follow the same convention as in https://github.com/uesteibar/elixir/blob/6c75771bf88a0ceda31cca25dfedd9c9636f7ac0/lib/iex/lib/iex/introspection.ex#L481, but I agree it's not the best name, I'll find a better one :)

**Comment by uesteibar** [comment]
@michalmuskala @eksperimental I addressed your feedback, thanks! 😸

File: lib/iex/lib/iex/introspection.ex [hunk 8]
```diff
@@ -115,13 +126,35 @@ defmodule IEx.Introspection do
...
       :no_docs
     end
   end
 
+  defp has_callback?(mod, fun) do
+    mod
+    |> Code.get_docs(:callback_docs)
+    |> find_callback_doc(fun)
+  end
+  defp has_callback?(mod, fun, arity) do
```
Comment (whatyouhide) [inline lib/iex/lib/iex/introspection.ex]: Let's leave a space between these two clauses. Same for other things like `find_callback_doc/2,3` below.

File: lib/iex/lib/iex/introspection.ex [hunk 9]
```diff
@@ -70,7 +72,14 @@ defmodule IEx.Introspection do
...
       end
 
-      if result != [], do: :ok, else: :not_found
+      cond do
+        result != []
+          -> :ok
+        has_callback?(mod, fun)
+          -> :behaviour_found
+        true
+          -> :not_found
```
Comment (whatyouhide) [inline lib/iex/lib/iex/introspection.ex]: The `->`s should go on the same line as the conditions:

```elixir
cond do
  result != [] ->
    :ok
  ...
end
```

Comment (uesteibar) [inline lib/iex/lib/iex/introspection.ex]: sorry, did it wrong when breaking the lines 🙈

**Comment by uesteibar** [comment]
@whatyouhide fixed :)

File: lib/iex/lib/iex/introspection.ex (hunk 4 above)
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: Yes, as i went on the review I noticed it was not introduced by you. Let's wait for others to give their input.

Comment (uesteibar) [inline lib/iex/lib/iex/introspection.ex]: I changed it here anyway since `for` is currently only used in the functions that take care of printing the final result, so `has_callback?` is now more consistent with the code that is around it.

File: lib/iex/lib/iex/introspection.ex [hunk 10]
```diff
@@ -115,13 +126,37 @@ defmodule IEx.Introspection do
         end
         :ok
       else
-        :not_found
+        if has_callback?(mod, fun, arity) do
+          :behaviour_found
+        else
+          :not_found
+        end
```
Comment (kelvinst) [inline lib/iex/lib/iex/introspection.ex]: Maybe you could refactor this chained `if`s to a `cond`, or even better, refactor it to multiple functions with pattern matching?

Comment (uesteibar) [inline lib/iex/lib/iex/introspection.ex]: Not a big friend of this nested `if` statements either, but it seems to be a common pattern on this file, so if we're refactoring this we could go for all of the `h` functions here.

If that's the case (and I'd love to do it), since we have more like *nested* `if`s than *chained*, I would rather go for something more like a [railway pattern](http://www.zohaib.me/railway-programming-pattern-in-elixir/) here, probably without the `>>>` thing and just handling the different scenarios through pattern matching.
This would make the different parts more reusable.

It would be something like:

```elixir
defp h_mod_fun_arity(mod, fun, arity) when is_atom(mod) do
    {mod, fun, arity}
    |> h_get_docs
    |> h_find_doc
    |> h_handle_doc
  end
```

I think it's a rather big topic to tackle on this PR, but I'd love to do it on another one.

What do you think? Let's see what the other folks here have to say :)

Comment (whatyouhide) [inline lib/iex/lib/iex/introspection.ex]: I think it's fine to do this on a different PR since this PR is in line with the rest of the file IMO. :)

Comment (uesteibar) [inline lib/iex/lib/iex/introspection.ex]: @whatyouhide awesome, I'll do it on another one then :)

All of the feedback I got on this one is addressed already.

File: lib/iex/lib/iex/introspection.ex [hunk 11]
```diff
@@ -115,13 +126,37 @@ defmodule IEx.Introspection do
...
+    mod
+    |> Code.get_docs(:callback_docs)
+    |> find_callback_doc(fun, arity)
+  end
+
+  def find_callback_doc(docs, fun) do
+    Enum.any?(docs, &match?({{^fun, _}, _, _, _}, &1))
+  end
+
+  def find_callback_doc(docs, fun, arity) do
```
Comment (whatyouhide) [inline lib/iex/lib/iex/introspection.ex]: `find_callback_doc/2,3` should be private, right? My personal opinion is that we could ditch these functions as we only use them once each and it would be clear enough but that's up to you and others in the team if they have a preference :).

Comment (uesteibar) [inline lib/iex/lib/iex/introspection.ex]: Ups, missed the `p` when cleaning it up...
I personally think it does make understanding the goal of the code easier, that's why I moved it to a separate function.

If other team members agree on not extracting it, it's fine to me. Meanwhile I'll fix the `defp` issue.

Thanks!

File: lib/iex/lib/iex/introspection.ex [hunk 12]
```diff
@@ -115,13 +126,37 @@ defmodule IEx.Introspection do
...
+    |> Code.get_docs(:callback_docs)
+    |> find_callback_doc(fun, arity)
+  end
+
+  defp find_callback_doc(docs, fun) do
+    Enum.any?(docs, &match?({{^fun, _}, _, _, _}, &1))
+  end
+
+  defp find_callback_doc(docs, fun, arity) do
+    Enum.any?(docs, &match?({{^fun, ^arity}, _, _, _}, &1))
```
Comment (josevalim) [inline lib/iex/lib/iex/introspection.ex]: Please use `List.keymember?`.

Comment (josevalim) [inline lib/iex/lib/iex/introspection.ex]: Actually, nevermind. The current way is fine because it is at least in sync with the line above.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
